packages/mutator-framework/src/mutation-node/model-property.ts

```diff
@@ -1,53 +1,66 @@
+import { HalfEdge } from "./mutation-edge.js";
 import { MutationNode } from "./mutation-node.js";
+import { traceNode } from "./tracer.js";
+
+export interface ModelPropertyConnectOptions {
+  /** Mutation key for the property's type node. Defaults to this node's key. */
+  typeMutationKey?: string;
+}
 
 export class ModelPropertyMutationNode extends MutationNode<ModelProperty> {
   readonly kind = "ModelProperty";
-  #referenceMutated = false;
-
-  traverse() {
-    const typeNode = this.subgraph.getNode(this.sourceType.type);
-    const referenceNode = this.subgraph.getReferenceNode(this.sourceType);
-
-    this.connectType(typeNode);
-    this.connectReference(referenceNode);
-  }
 
-  connectReference(referenceNode: MutationNode<Type>) {
-    MutationEdge.create(this, referenceNode, {
-      onTailMutation: () => {
-        this.#referenceMutated = true;
-        this.mutatedType.type = referenceNode.mutatedType;
+  startTypeEdge() {
+    return new HalfEdge<ModelProperty, Type>(this, {
+      onTailMutation: (tail) => {
+        traceNode(this, "Model property type mutated.");
+        this.mutate();
+        this.mutatedType.type = tail.mutatedType;
       },
       onTailDeletion: () => {
-        this.#referenceMutated = true;
+        this.mutate();
         this.mutatedType.type = this.$.intrinsic.any;
       },
-      onTailReplaced: (newTail) => {
-        this.#referenceMutated = true;
-        this.mutatedType.type = newTail.mutatedType;
+      onTailReplaced: (_oldTail, newTail, head, reconnect) => {
+        head.mutate();
+        head.mutatedType.type = newTail.mutatedType;
+        if (reconnect) {
+          (head as ModelPropertyMutationNode).connectType(newTail);
+        }
+      },
+      onHeadReplaced: (_oldHead, newHead, tail) => {
+        // When this edge's head is replaced, have the new head establish
+        // its own connection to the tail so it receives future mutations
+        if (newHead instanceof SelfType) {
+          (newHead as ModelPropertyMutationNode).connectType(tail);
+        }
       },
     });
   }
 
   connectType(typeNode: MutationNode<Type>) {
-    MutationEdge.create(this, typeNode, {
-      onTailMutation: () => {
-        if (this.#referenceMutated) {
-          return;
-        }
-        this.mutatedType.type = typeNode.mutatedType;
+    this.startTypeEdge().setTail(typeNode);
+  }
+
+  startModelEdge() {
+    return new HalfEdge<ModelProperty, Model>(this, {
+      onTailMutation: (tail) => {
+        traceNode(this, "Model property model mutated.");
+        this.mutate();
+        this.mutatedType.model = tail.mutatedType;
       },
       onTailDeletion: () => {
-        if (this.#referenceMutated) {
-          return;
-        }
-        this.mutatedType.type = this.$.intrinsic.any;
+        this.delete();
       },
-      onTailReplaced: (newTail) => {
-        this.mutatedType.type = newTail.mutatedType;
+      onTailReplaced: (_oldTail, _newTail, head, _reconnect) => {
```

Comment:
seeing a few which need to have `_` to ignore the params wouldn't this signature be better with an object bag?

Reply:
I dunno, not really an options bag per se since everything is always present, but I'll make that change.